### PR TITLE
[NFC][AlwaysInliner] Reduce AlwaysInliner memory consumption.

### PR DESCRIPTION
**Pull request by dfukalov**
Refactored AlwaysInliner to remove some of inlined functions earlier.

Before the change AlwaysInliner walked through all functions in the module and inlined them into calls where it is appropriate. Removing of the dead inlined functions was performed only after all of inlining. For the test case from the issue [59126](https://github.com/llvm/llvm-project/issues/59126) compiler consumes all of the memory on 64GB machine, so is killed.

The change checks if just inlined function can be removed from the module and removes it.

### CODE REVIEW

File: llvm/test/Transforms/Inline/always-inline-phase-ordering.ll [hunk 1]
```diff
@@ -1,12 +1,10 @@
 ; RUN: opt --Os -pass-remarks=inline -S < %s 2>&1 | FileCheck %s
 target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
-target triple = "arm64e-apple-macosx13"
 
 ; CHECK: remark: <unknown>:0:0: 'wibble' inlined into 'bar.8' with (cost=always): always inline attribute
 ; CHECK: remark: <unknown>:0:0: 'wibble' inlined into 'pluto' with (cost=always): always inline attribute
 ; CHECK: remark: <unknown>:0:0: 'snork' inlined into 'blam' with (cost=always): always inline attribute
 ; CHECK: remark: <unknown>:0:0: 'wobble' inlined into 'blam' with (cost=always): always inline attribute
-; CHECK: remark: <unknown>:0:0: 'wobble' inlined into 'snork' with (cost=always): always inline attribute
```
Comment (dfukalov) [inline llvm/test/Transforms/Inline/always-inline-phase-ordering.ll:9]: This remark was about inlining 'wobble' into already inlined 'snork'.

File: llvm/test/Transforms/Inline/always-inline-phase-ordering.ll [hunk 2]
```diff
@@ -1,12 +1,10 @@
 ; RUN: opt --Os -pass-remarks=inline -S < %s 2>&1 | FileCheck %s
 target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
-target triple = "arm64e-apple-macosx13"
```
Comment (aeubanks) [inline llvm/test/Transforms/Inline/always-inline-phase-ordering.ll:3]: please split the unrelated test changes to another PR

**Comment by aeubanks** [comment]
can you elaborate in the description on "Refactored AlwaysInliner to remove some of inlined functions earlier" and how it fixes the issue?

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp [hunk 3]
```diff
@@ -38,6 +38,7 @@ bool AlwaysInlineImpl(
   SmallSetVector<CallBase *, 16> Calls;
   bool Changed = false;
   SmallVector<Function *, 16> InlinedFunctions;
+  SmallVector<Function *, 16> WorkList;
   for (Function &F : M) {
```
Comment (aeubanks) [inline llvm/lib/Transforms/IPO/AlwaysInliner.cpp:42]: I think we can do `for (Function &F: make_early_inc_range(M))` to enable deleting in this loop

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp [hunk 4]
```diff
@@ -46,64 +47,73 @@ bool AlwaysInlineImpl(
...
+        M.getFunctionList().erase(F);
+        Changed = true;
       }
     }
   }
 
-  // Remove any live functions.
+  // Final cleanup stage. Firstly, remove any live functions.
   erase_if(InlinedFunctions, [&](Function *F) {
     F->removeDeadConstantUsers();
```
Comment (aeubanks) [inline llvm/lib/Transforms/IPO/AlwaysInliner.cpp:111]: I don't think we need this part anymore if we've already called it on every function in the loop above?

**Review by aeubanks — COMMENTED** [review]
so the issue is that if we have for example a bunch of internal functions that are used once by another internal function, we may run into quadratic memory usage if everything is inlined? that makes sense

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp [hunk 5]
```diff
@@ -46,64 +47,73 @@ bool AlwaysInlineImpl(
...
+      emitInlinedIntoBasedOnCost(
+          ORE, DLoc, Block, *F, *Caller,
+          InlineCost::getAlways("always inline attribute"),
+          /*ForProfileContext=*/false, DEBUG_TYPE);
+
+      Changed = true;
+    }
+
+    F->removeDeadConstantUsers();
+    if (F->hasFnAttribute(Attribute::AlwaysInline) && F->isDefTriviallyDead()) {
```
[the remark this answers is not in the document]
Comment (dfukalov) [reply llvm/lib/Transforms/IPO/AlwaysInliner.cpp:96]: without checking `AlwaysInline` attribute the change will be not NFC: it will process here not only `AlwaysInline` functions, but also can remove some other dead functions. I'm not sure it is intended in the pass.

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp [hunk 6]
```diff
@@ -46,64 +47,73 @@ bool AlwaysInlineImpl(
...
+      // Remember to try and delete this function afterward. This both avoids
+      // re-walking the rest of the module and avoids dealing with any
+      // iterator invalidation issues while deleting functions.
+      if (F->hasComdat()){
```
[the remark this answers is not in the document]
Comment (dfukalov) [reply llvm/lib/Transforms/IPO/AlwaysInliner.cpp:100]: thanks!

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp (hunk 3 above)
Comment (dfukalov) [reply llvm/lib/Transforms/IPO/AlwaysInliner.cpp:42]: yes, thanks!

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp (hunk 4 above)
Comment (dfukalov) [reply llvm/lib/Transforms/IPO/AlwaysInliner.cpp]: Yes, you're right, thanks, removed this part.

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp [hunk 7]
```diff
@@ -38,73 +38,70 @@ bool AlwaysInlineImpl(
...
-    F->removeDeadConstantUsers();
-    return !F->isDefTriviallyDead();
-  });
-
   // Delete the non-comdat ones from the module and also from our vector.
-  auto NonComdatBegin = partition(
-      InlinedFunctions, [&](Function *F) { return F->hasComdat(); });
+  auto *NonComdatBegin =
+      partition(InlinedFunctions, [&](Function *F) { return F->hasComdat(); });
   for (Function *F : make_range(NonComdatBegin, InlinedFunctions.end())) {
```
Comment (aeubanks) [inline llvm/lib/Transforms/IPO/AlwaysInliner.cpp:105]: this part is obsolete now right? we don't have any non-comdat functions in `InlinedFunctions` (should also change the variable name to `InlinedComdatFunctions`

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (aeubanks) [reply llvm/lib/Transforms/IPO/AlwaysInliner.cpp]: fair enough

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp (hunk 7 above)
Comment (dfukalov) [reply llvm/lib/Transforms/IPO/AlwaysInliner.cpp]: yes, thanks, removed

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp [hunk 8]
```diff
@@ -37,86 +37,74 @@ bool AlwaysInlineImpl(
...
   bool Changed = false;
-  SmallVector<Function *, 16> InlinedFunctions;
-  for (Function &F : M) {
-    // When callee coroutine function is inlined into caller coroutine function
-    // before coro-split pass,
-    // coro-early pass can not handle this quiet well.
-    // So we won't inline the coroutine function if it have not been unsplited
+  SmallVector<Function *, 16> InlinedComdatFunctions;
+
+  for (Function &F : make_early_inc_range(M)) {
```
Comment (aeubanks) [inline llvm/lib/Transforms/IPO/AlwaysInliner.cpp:42]: add a comment about the fact that we need this since we may delete the function we're visiting

**Review by aeubanks — APPROVED** [review]
lgtm with comment added, thanks for doing this

File: llvm/lib/Transforms/IPO/AlwaysInliner.cpp (hunk 8 above)
Comment (dfukalov) [reply llvm/lib/Transforms/IPO/AlwaysInliner.cpp:42]: I've checked a number of places `make_early_inc_range()` used - found no comments about its functionality. It seems its usage instead of default iterator is quite obvious, what do you think?

Updated the comment below about `InlinedComdatFunctions` usage.

Comment (aeubanks) [reply llvm/lib/Transforms/IPO/AlwaysInliner.cpp:42]: I think it's not immediately obvious since the deletion happens further down so I'd write a comment, but won't object if you omit the comment
